### PR TITLE
[ray_client]: Insert decorators into the real ray module to allow for client mode

### LINKED ISSUE #12423: Get Ray Tune / RLlib workload to run in client mode [5]
This concretely means tune.run(PPOTrainer) works (printing out logs etc, similar to in a normal Ray cluster).

### LINKED ISSUE #12983: Excise the trampoline from Ray Client [3]
### Describe your feature request

After #12818 and #12964 the client/server pickling convention means that the ray object itself can be stubbed.

This cleaner approach means deleting a lot of code -- there's no need for the elaborate CoreRayAPI anymore, it's merely the client Ray API.

### PR DESCRIPTION
**Pull request by barakmich**
How it works:
```
RAY_CLIENT_MODE=1 python <everyday ray file>
```

To wit, 
```
RAY_CLIENT_MODE=1 python python/ray/tune/examples/pbt_example.py
```
works unmodified.

Also, align the tests which already had a flag to use the new flag.

## Why are these changes needed?

This finally allows exposure of the experimental ray client without modifying (much) ray code.

## Related issue number

Closes #12423 

Closes #12983

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/_private/client_mode_hook.py [hunk 1]
```diff
@@ -0,0 +1,45 @@
+import os
+from contextlib import contextmanager
+from ray.experimental.client import ray
+
+client_mode_enabled = os.environ.get("RAY_CLIENT_MODE", "0") == "1"
```
Comment (ericl) [inline python/ray/_private/client_mode_hook.py:5]: Is the reason for an environment variable to handle remote function definitions that occur before `ray.connect()`?

We should be able to handle this in the future so you don't need to set an env var.

File: python/ray/experimental/client/server/server.py [hunk 2]
```diff
@@ -17,35 +17,34 @@
...
-        self._test_mode = test_mode
         self._current_function_stub = None
 
     def ClusterInfo(self, request,
                     context=None) -> ray_client_pb2.ClusterInfoResponse:
         resp = ray_client_pb2.ClusterInfoResponse()
         resp.type = request.type
         if request.type == ray_client_pb2.ClusterInfoType.CLUSTER_RESOURCES:
-            resources = ray.cluster_resources()
+            with disable_client_hook():
```
Comment (ericl) [inline python/ray/experimental/client/server/server.py:46]: I'm a bit confused why this is needed--- isn't RAY_CLIENT_ENABLED False on the server?

File: python/ray/_raylet.pyx [hunk 3]
```diff
@@ -558,6 +562,7 @@ cdef CRayStatus task_execution_handler(
 
     with gil:
         try:
+            client_was_enabled = _disable_client_hook()
```
Comment (ericl) [inline python/ray/_raylet.pyx:565]: When is this needed? Won't the client always be disabled on the server?

**Review by ericl — CHANGES_REQUESTED** [review]

File: python/ray/worker.py [hunk 4]
```diff
@@ -468,6 +469,7 @@ def get_dashboard_url():
 """ray.node.Node: The global node object that is created by ray.init()."""
 
 
+@client_mode_hook
 def init(
```
Comment (ericl) [inline python/ray/worker.py:473]: Shouldn't the user be required to use `ray.connect()`?

File: python/ray/worker.py [hunk 5]
```diff
@@ -780,6 +782,7 @@ def init(
 _post_init_hooks = []
 
 
+@client_mode_hook
 def shutdown(_exiting_interpreter=False):
```
Comment (ericl) [inline python/ray/worker.py:786]: Shouldn't shutdown be disallowed?

**Comment by ericl** [comment]
Ok I see, the env var is a way to run a single script in emulated client mode execution. While useful, this isn't exactly the goal of https://github.com/ray-project/ray/issues/12423, which is to test that *real* client mode execution works with Tune.

It seems like the hook gets you 90% of the way there, but when I run
```
from ray.experimental.client import ray
from ray.tune import tune

ray.connect("localhost:50051")

tune.run("PG", config={"env": "CartPole-v0"})
```

I see that actually it starts a second instance of Ray since the `ray.is_initialized()` call isn't properly returning True after `ray.connect`. Probably you need to set the client mode = True after `ray.connect` at least.

File: python/ray/_private/client_mode_hook.py (hunk 1 above)
Comment (barakmich) [reply python/ray/_private/client_mode_hook.py:5]: Yep, that's it. We need a way to make sure that on startup all the calls are hooked.

I'm not sure there's a great future option -- we'd have to convert or genericize the remote functions and actors, which requires some design.

For a MVP feature, a flag works surprisingly well.

File: python/ray/experimental/client/server/server.py (hunk 2 above)
Comment (barakmich) [reply python/ray/experimental/client/server/server.py:46]: If you're running a server in a separate process than the client, sure. This becomes a noop. 

But in tests and in cases where it's trying to emulate `ray.init()` by way of running both locally, it's a bit more complicated.

That is an alternative for a follow-up -- spawn the server as a new process and handle that, with the right env changes -- but that's not how we've been running this whole time.

File: python/ray/_raylet.pyx (hunk 3 above)
Comment (barakmich) [reply python/ray/_raylet.pyx:565]: Same as above -- especially in local mode.

In fact, local mode is exactly the reason to consider this approach. Local mode expects much more about running in process.

File: python/ray/worker.py (hunk 4 above)
Comment (barakmich) [reply python/ray/worker.py:473]: That's a valid design option. The nice thing about doing this is, again, everywhere (including tests and examples and general hands-on experience) where folks expect `init()` to work locally, this provides that.

File: python/ray/worker.py (hunk 5 above)
Comment (barakmich) [reply python/ray/worker.py:786]: Ditto above, with `init()`

**Comment by barakmich** [comment]
> which is to test that _real_ client mode execution works with Tune.

Which, I mean, if it's going through all the same code paths, is no different than the tests. However, the following is totally valid:

> 
> It seems like the hook gets you 90% of the way there, but when I run
> 
> ```
> from ray.experimental.client import ray
> from ray.tune import tune
> 
> ray.connect("localhost:50051")
> 
> tune.run("PG", config={"env": "CartPole-v0"})
> ```

I think this is totally reasonable to work, sans env var. Will give it a shot.

**Comment by barakmich** [comment]
I think the underlying thing I wanted to achieve -- env var and all -- was that you specifically have to opt in to get the client functionality. Even the decorator is a noop if it's not set (which I'm needing to change to make the explicit-client-import bit to work). My philosophy was to be as minimally invasive as possible, but that may not be the desired behavior 🤷

**Comment by barakmich** [comment]
Here we go.
Window 1:
```
python ray/experimental/client/server/server.py
```

Window 2:
```
python ray/experimental/client/examples/run_tune.py
```

Sure enough, training occurs on the server side with run_tune driving it.
run_tune.py being exactly the code provided.

**Review by ericl — APPROVED** [review]
:+1: works as advertised

File: python/ray/experimental/client/__init__.py [hunk 6]
```diff
@@ -1,148 +1,108 @@
...
+        self._inside_client_test = False
 
-class RayAPIStub:
     def connect(self,
                 conn_str: str,
                 secure: bool = False,
-                metadata: List[Tuple[str, str]] = None,
-                stub=None) -> None:
+                metadata: List[Tuple[str, str]] = None) -> None:
+        """
```
Comment (ericl) [inline python/ray/experimental/client/__init__.py:29]: Style nits: no newline (`"""Connect the Ray Client to a server`) here and elsewhere.

**Comment by ericl** [comment]
Will merge pending tests.

**Comment by ericl** [comment]
Some interesting sphinx errors in lint: https://travis-ci.com/github/ray-project/ray/jobs/464169979

Perhaps import order related?

**Comment by ericl** [comment]
We can probably solve by making client or pb2 imports lazy. Seems sphinx can't access the genfiles

**Comment by barakmich** [comment]
Woof. Yeah, I've got an idea

**Comment by ericl** [comment]
No dice

**Comment by barakmich** [comment]
LINT=1 doesn't build anything like it usually should. there's an intermediate target (I actually created in one of these client prs) for just the python protos... so that last commit should work one way or another

**Comment by ericl** [comment]
Still no good. Try sphinx building locally?

**Comment by barakmich** [comment]
yeah, i'm looking into it

**Comment by barakmich** [comment]
🎉

**Pull request merged**